### PR TITLE
handler, s3store: Serve GET requests directly from S3

### LINKED ISSUE #1064: handler: support HTTP range requests in GET /:upload
Some discussion around API design in the quoted (since-closed) PR.

>> Out of interest, I did some searching and found #432 where you were against implementing `Range` support in tusd - have you changed your mind on this?

>Yes, I did change my opinion here. We have recently been involved in the HTTP working group with the goal of making a IETF standard for resumable uploads. In the end, we hope that resumable uploads are one component in the HTTP toolbox that interacts well with the existing HTTP mechanism. Since then I am more interested in making sure that tusd works better with concepts, such as range requests.

>> In any case, we should probably move this conversation to an issue to sketch out API design if you're in favour of this in principle - closing this PR for now.

>Great idea, feel free to open an issue if you are still interested in working on this :)

_Originally posted by @Acconut in https://github.com/tus/tusd/issues/1048#issuecomment-1881180105_

### PR DESCRIPTION
**Pull request by pcfreak30**
- Add ContentServerDataStore interface
- Update handlers to use ContentServerDataStore when available
- Implement range request handling for s3upload
- Add tests for new ContentServerDataStore functionality
- Update Go version to 1.22.1

Closes #1064

### CODE REVIEW

**Comment by pcfreak30** [comment]
Please note I tried to avoid changing the go version, but `go mod tidy` seems to automatically do it.

This is a draft to get feedback.

@Acconut

File: pkg/s3store/s3store.go
Comment (Acconut) [inline pkg/s3store/s3store.go:107]: If we implement `ServeContent` directly on `s3Upload`, we do not have to introduce another struct `S3ServableUpload`. This is already what we do for `AsTerminatableUpload`:

https://github.com/tus/tusd/blob/9779a84cae29d083b3496ba732fe7c76755bd4ed/pkg/s3store/s3store.go#L367-L369

File: pkg/s3store/s3store.go
Comment (Acconut) [inline pkg/s3store/s3store.go:118]: Before merging, we should move these functions down in the file as they are not the most important ones. Maybe place the logic next to the other optional interfaces that S3Store implements.

File: pkg/s3store/s3store.go
Comment (Acconut) [inline pkg/s3store/s3store.go:1339]: Can you explain what the benefit of this is over `io.Copy`?

File: pkg/s3store/s3store.go
Comment (Acconut) [inline pkg/s3store/s3store.go:157]: I'm wonder whether dedicated handling for range requests is necessary. If the request includes the Range header we can add it to `GetObject` without parsing the ranges and let S3 deal with it. Afterwards we just copy the Content-Range, Content-Length and ETag header from S3's response to tusd's response and copy the body.

The advantage is that we don't have to parse the header (and possibly introduce subtle bugs) and don't have to construct the Content-Range and Content-Length headers on our own.

We would also not need a dedicated `handleRangeRequest` function as this could all live in `ServeContent`.

What do you think?

File: pkg/handler/composer.go
Comment (Acconut) [inline pkg/handler/composer.go:17]: As mentioned in https://github.com/tus/tusd/issues/1064#issuecomment-2401568772, I am not sure on my own whether `Server` is a good name as it does have a different meaning in the HTTP-land. How about `ContentServer` as an alternative?

We would then have `StoreComposer.ContentServer`, `StoreComposer.UsesContentServer`, `StoreComposer.UseContentServer`, and `tusd.ContentServerDataStore`.

**Review by Acconut — COMMENTED** [review]
Thank you very much for working on this! I left a few comments but the direction where this is heading is looking good.

File: pkg/s3store/s3store.go
Comment (pcfreak30) [reply pkg/s3store/s3store.go:107]: I can change this, though I did see value in it.

File: pkg/s3store/s3store.go
Comment (pcfreak30) [reply pkg/s3store/s3store.go:118]: :+1:

File: pkg/s3store/s3store.go
Comment (pcfreak30) [reply pkg/s3store/s3store.go:1339]: TBH, I rely on AI for development and a lot of this I was able to complete due to its effort. The data copy and range logic is the most complicated parts, and as I did not have a good reason to remove them based on what I saw, I just ensured it was all valid.

If needed, I can replace with io.Copy and move on.

File: pkg/s3store/s3store.go
Comment (pcfreak30) [reply pkg/s3store/s3store.go:157]: I believe given that multiple ranges are possible, it made since to check it ourselves. and due to the complexity of the range checking, its why I had a unit test for the range logic made.

I can punt it all to be forwarded to s3, but I also think it is valid to verify at-least the ranges earlier in  the s3 impl.

File: pkg/handler/composer.go
Comment (pcfreak30) [reply pkg/handler/composer.go:17]: works for me, I don't really care :P.

**Comment by pcfreak30** [comment]
@Acconut changes have been made. kudos.

File: pkg/handler/datastore.go
Comment (Acconut) [inline pkg/handler/datastore.go:127]: This function needs proper documentation. That's best done when we have agreed on what the handler does and does not.

File: pkg/handler/unrouted_handler.go
Comment (Acconut) [inline pkg/handler/unrouted_handler.go:1019]: I wonder if the handler should set the Content-Length, Content-Type and Content-Disposition before passing it to `ServeContent`. The implementation can then decide to overwrite or keep them. The idea is that we already have logic for extracting the filename and filetype from metadata and this way we can keep using them.

File: pkg/s3store/s3store.go
Comment (Acconut) [reply pkg/s3store/s3store.go]: The goal of the new ServeContent interface is to reduce the complexity that we have to handle inside of tusd and adding logic to each store to validate the Range header is not having the opposite effect. In addition, deferring the validation to S3 (or whatever provider other stores use) is preferable since it knows best which values it will accept and we avoid discrepancies between tusd's validation and S3's.

File: pkg/s3store/s3store.go
Comment (Acconut) [inline pkg/s3store/s3store.go:416]: How about forwarding the Content-Disposition and Content-Encoding headers as well if they are present? I think that would be useful as well.

**Review by Acconut — CHANGES_REQUESTED** [review]
Thank you for the updates! This is moving in the right direction and I feel like we're getting close :)

File: pkg/s3store/s3store.go
Comment (pcfreak30) [reply pkg/s3store/s3store.go:416]: done

File: pkg/handler/unrouted_handler.go
Comment (pcfreak30) [reply pkg/handler/unrouted_handler.go:1053]: you decide?

File: pkg/handler/datastore.go
Comment (pcfreak30) [reply pkg/handler/datastore.go:127]: lmk what/when you want docs on this.

**Comment by pcfreak30** [comment]
@Acconut changes have been made.

File: pkg/s3store/s3store.go
Comment (pcfreak30) [reply pkg/s3store/s3store.go]: done

File: pkg/handler/unrouted_handler.go
Comment (Acconut) [reply pkg/handler/unrouted_handler.go:1053]: Just implemented this by myself.

Comment (pcfreak30) [reply pkg/handler/unrouted_handler.go:1053]: Sorry, huh? where?

**Comment by Acconut** [comment]
Thank you for the updates! I added some documentation and better handling for range requests and conditional requests, in particular regarding error responses. I couldn't push these changes to this PR because you didn't allow maintainers to modify your PR. So for now, I put them into a separate branch: https://github.com/tus/tusd/compare/main...s3-content-server

Feel free to review them. If you allow me to modify this PR, I can push them here. If not, we'll find another way to collaborate.

**Comment by pcfreak30** [comment]
@Acconut sorry GitHub doesn't seem to be allowing me to give you access :/.

**Comment by Acconut** [comment]
Ok, then I'll continue the work in my own PR: https://github.com/tus/tusd/pull/1228
Thank you for your help so far!

**Comment by Acconut** [comment]
I just merged this as part of https://github.com/tus/tusd/pull/1228. Thank you!

**Pull request closed without merging**
